src/coreclr/jit/emit.cpp

```diff
@@ -4479,6 +4610,429 @@ void emitter::emitJumpDistBind()
 #endif // DEBUG
 }
 
+#ifdef FEATURE_LOOP_ALIGN
+
+void emitter::emitLoopAlignment()
+{
+    if ((emitComp->opts.compJitAlignLoopBoundary > 16) && (!emitComp->opts.compJitAlignLoopAdaptive))
+    {
+        emitLongLoopAlign(emitComp->opts.compJitAlignLoopBoundary);
+    }
+    else
+    {
+        emitLoopAlign();
+    }
+
+    // Mark this IG as need alignment so during emitter we can check the instruction count heuristics of
+    // all IGs that follows this IG and participate in a loop.
+    emitCurIG->igFlags |= IGF_LOOP_ALIGN;
+
+    JITDUMP("Adding 'align' instruction of %d bytes in G_M%03u_IG%02u.\n", emitComp->opts.compJitAlignLoopBoundary,
+            emitComp->compMethodID, emitCurIG->igNum);
+}
+
+//-----------------------------------------------------------------------------
+//  For loopHeaderIg, find the size of the smallest possible loop that doesn't exceed maxLoopSize.
+//
+unsigned emitter::getLoopSize(insGroup* igLoopHeader, unsigned maxLoopSize)
+{
+    unsigned loopSize = 0;
+
+    for (insGroup* igInLoop = igLoopHeader; igInLoop != nullptr; igInLoop = igInLoop->igNext)
+    {
+        loopSize += igInLoop->igSize;
+        if (igInLoop->isLoopAlign())
+        {
+            // If igInLoop's next IG is a loop and needs alignment, then igInLoop should be the last IG
+            // of the current loop and should have backedge to current loop header.
+            assert(igInLoop->igLoopBackEdge == igLoopHeader);
+
+            // In such cases, the current loop size should exclude the align instruction size reserved for
+            // next loop.
+            unsigned maxPaddingAllowed;
+            if (emitComp->opts.compJitAlignLoopAdaptive)
+            {
+                maxPaddingAllowed = (emitComp->opts.compJitAlignLoopBoundary >> 1) - 1;
+            }
+            else
+            {
+                maxPaddingAllowed = emitComp->opts.compJitAlignLoopBoundary - 1;
+            }
+
+            loopSize -= maxPaddingAllowed;
+        }
+        if ((igInLoop->igLoopBackEdge == igLoopHeader) || (loopSize > maxLoopSize))
+        {
+            break;
+        }
+    }
+
+    return loopSize;
+}
+
+//-----------------------------------------------------------------------------
+// emitSetLoopBackEdge : Sets igLoopBackEdge field, if not already set and
+//                       if currIG has back-edge to dstIG.
+//
+// Notes:
+//    If the current loop covers a loop that is already marked as align, then remove
```

Comment:
Can you describe this differently?
```suggestion
//    If the current loop encloses a loop that is already marked as align, then remove
```
Also what happens if the two loops intersect but one doesn't enclose the other?

Reply:
Could you give an example?